feat: add initial p2tr output script support

sangaman opened the pull request:
This introduces support for creating p2tr output scripts and addresses from one or more pubkeys and/or one or more scripts according to BIPs 340 & 341.

Multiple pubkeys are aggregated according to the procedure defined by MuSig whereby all keys are concatenated and hashed, with the resulting hash multiplied against each pubkey to tweak it. These tweaked pub keys are added together to produce the internal pubkey. If only one pubkey is provided, then the internal pubkey is simply set to that. And if no pubkeys are provided, a key with no known secret key is used instead.

Any scripts provided are arranged into a Huffman tree according to their weighted probability and then hashed to form a merkle tree. The root hash of this tree is then added to the internal pubkey to tweak it. This tweaked key is finally used for creating the output script which is then encoded using Bech32m.

Test cases in this commit include examples from the Bitcoin Optech taproot workshop that derive a Bech32m address from scripts and an internal key.

Follow-ups to this commit should include expanding the coverage and number of test cases as well as possibly restructuring existing code or moving logic to other packages/modules. We should also consider support for taproot output descriptors, which are currently under review and development in Bitcoin Core.

Ticket: BG-35571

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+ */
+export function aggregateMuSigPubkeys(pubkeys: Buffer[]): Buffer {
+  // sort keys in ascending order
+  pubkeys.sort();
+
+  const trimmedPubkeys: Buffer[] = [];
+  pubkeys.forEach(pubkey => {
+    const trimmedPubkey = trimFirstByte(pubkey);
+    trimmedPubkeys.push(trimmedPubkey);
+  });
```
```suggestion
const trimmedPubkeys = pubkeys.map(trimFirstByte).sort();
```
I _think_ the byte we're trimming could effect sort order, but maybe not if bitcoin always uses the odd|even pubkey. Either way, this is cleaner IMO.

sangaman replied:
Very good point, this ought to be sorting only on the 32 byte x coordinate, ignoring any leading `0x02`/`0x03` byte. Plus this is just more succinct.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+
+  const L = bcrypto.sha256(Buffer.concat(trimmedPubkeys));
+
+  let aggregatePubkey: Buffer;
+  pubkeys.forEach(pubkey => {
+    const trimmedPubkey = trimFirstByte(pubkey);
+    const c = bcrypto.sha256(Buffer.concat([L, trimmedPubkey]));
+
+    const tweakedPubkey = ecc.pointMultiply(pubkey, c);
+    if (aggregatePubkey === undefined) {
+      aggregatePubkey = tweakedPubkey;
+    } else {
+      aggregatePubkey = ecc.pointAdd(aggregatePubkey, tweakedPubkey);
+    }
+  });
```
```suggestion
const tweakedPubkeys = pubkeys .map((pubkey) => {
  const trimmedPubkey = trimFirstByte(pubkey);
  const c = bcrypto.sha256(Buffer.concat([L, trimmedPubkey]));
  return ecc.pointMultiply(pubkey, c);
})
const aggregatePubkey = tweakedPubkeys.slice(1).reduce(ecc.pointAdd, tweakedPubkeys[0]);
```
Alternate construction - kinda two separate ideas:
1) make tweaking and combining the pubkeys separate, because this helps show that order doesn't matter
2) use a slightly clever `reduce` construction to eliminate the ugly conditional

sangaman replied:
I like it. Needed to tweak this slightly though, the suggested code strangely caused the tests to fail. Turns out that `pointAdd` takes a third parameter indicating whether the key is compressed or not, if falsey it returns a 65 byte uncompressed key instead of the 33 byte compressed key. The third parameter for `reduce` is the current index so for index `0`, `pointAdd` treats that as false and returns a 65 byte key and messes the whole thing up. Got it to work with `reduce((prev, curr) => ecc.pointAdd(prev, curr), tweakedPubkeys[0]);`. Had me questioning my sanity briefly though.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+
+    tweakedPubkey = ecc.pointAddScalar(pubkey, tapTweak);
+  } else {
+    // If the spending conditions do not require a script path, the output key should commit to an
```
Can you link me to the explanation of this?

sangaman replied:
https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-22 - I'll add this to the comments.

sangaman replied:
Looking at this again I was definitely missing the `hashTapTweak` tagged hash part. I *believe* though that the multiplication by the generator gets handled within the `pointAddScalar` function, although I may be wrong. I wouldn't be sure how to do it otherwise. Getting my hands on a test case would be ideal, as I could pin down this logic pretty easily with that. No luck so far though.

sangaman replied:
Following up on this the way I currently have it with the taptweak tagged hash makes sense to me at least. We're tweaking the internal pubkey with a tagged hash just like we do when we have an actual script path, thus not leaking that there is no script path, but we're doing it with something that can't be used for script spending.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+
+    tweakedPubkey = ecc.pointAddScalar(pubkey, unspendableScriptPathRoot);
+  }
+
```
nit: Might be nicer to shrink the conditional block by doing `let tapTweak;` at the top, and then doing the `pointAddScalar` right before return.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+ * @param weights
+ * @returns the tagged hash of the taptree root
+ */
+export function getHuffmanTaptreeRoot(
```
This function is a tiny bit hard to reason about, but I love it.

sangaman replied:
Yep I struggled a bit with how to frame/structure this. It does quite a bit, but I also figure we'll need more from this construction eventually because we need to preserve knowledge of the internal hashes of the tree to be able to spend using scripts. For now it just constructs the tree and gives us the root hash, which is all we need for creating the output script. I'll go through this again and add some comments, but it will need to be reworked eventually either way I'm pretty sure.

tylerlevine replied:
Yes I find it quite nice too - some more comments might be useful though, especially around the reasoning on why the non-null assertions are safe (which I believe them to be).

tylerlevine reviewed (changes requested):
excellent work - this is a really great PR here!

No major blockers but I think some additional test cases that cross reference test vectors from other p2tr or musig2 key aggregation implementations would be useful.

tylerlevine commented:
also please check the failing lint step - just looks like the imports need to be ordered

sangaman commented on test/fixtures/p2tr.json:
```diff
@@ -0,0 +1,70 @@
+{
```
The first two cases are from the optech taproot workshop (specifically section 2.4). The first one is basically lifted straight from one of their examples - it uses the same internal pubkey but the scripts are generated from random keys (1 per script). For the second one I tweaked the python code to create an aggregated pubkey from 2 of the 3 keys, and tweaked the scripts to be 2 of 2 multisigs using `OP_CHECKSIGVERIFY` (not that the correctness of the scripts matters per se, but I was trying to approximate the BitGo multisig use case). In both cases I'm checking that we get the same outputs as with the optech workshop code.

The second two are just taken from scripts/addresses I randomly generated with Bitcoin Core using https://github.com/bitcoin/bitcoin/pull/22364 - just double-checking the output script to address functionality.

sangaman commented on test/taproot.spec.ts:
```diff
@@ -0,0 +1,110 @@
+import * as assert from 'assert';
+import { describe, it } from 'mocha';
```
Indeed, this was a copy-paste job from another test suite. Looks like a lot of the existing ones import this unnecessarily, fwiw.

sangaman commented on test/taproot.spec.ts:
```diff
@@ -0,0 +1,110 @@
+import * as bcrypto from '../src/crypto';
+import * as taproot from '../src/taproot';
+import { ECPair } from '..';
+export const OPS = require('bitcoin-ops') as { [index: string]: number };
```
This was another copy-paste job, exporting is totally unnecessary here. Trying to import and use types here instead of `require`ing results in `This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag` so I won't bother it.

Since I'm only using it once in this spec file, and the specific op code isn't particularly important to the test, I'll just use the raw value and add a comment.

sangaman commented on test/taproot.spec.ts:
```diff
@@ -0,0 +1,110 @@
+    'hex',
+  );
+
+  it('aggregates pubkeys', () => {
```
I'll give that a shot as I agree more thorough test cases are needed. I did look there but couldn't come up with anything I could just grab and use here for the output side of things only. Running tests w/ debug statements could work tho at least for the aggregated key side of things, I'd still want more coverage for taptree creation and key tweaking.

sangaman commented on ts_src/payments/p2tr.ts:
```diff
@@ -0,0 +1,84 @@
+const { bech32m } = require('bech32');
+
+/** Internal key with unknown discrete logarithm for eliminating keypath spends */
+const H = Buffer.from(
```
I've updated the comment for this variable with some more detail about how this value is derived. I went with `H` because that's what used in BIP341 to describe this variable. Personally I think it's OK since it's only used in one place and that has comments for what it's for, along with the comments for the variable here. I think unless you're familiar with that section of BIP341 even a more descriptive name like `UNKNOWN_DLOG_INTERNAL_KEY` will be unclear without context.

sangaman commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+import * as bcrypto from './crypto';
+const ecc = require('tiny-secp256k1');
+
+// const SECP256K1_ORDER = Buffer.from('fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141', 'hex')
```
Yeah, I was originally going to use this for manually negating keys but I don't think that's necessary here.

sangaman commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+  // _MuSig coefficient_ computed according to the following equation
+  //
+  // L = H(P_1 || P_2 || ... || P_n)
+  // µ_i = H(L || i)
```
You're absolutely right... I actually copied this line (since there's no µ key on my keyboard) from the musig readme in secp256k1-zkp here https://github.com/ElementsProject/secp256k1-zkp/blob/master/src/modules/musig/musig.md. It's probably worth fixing there as well, you can do the honors if you'd like since you found the mistake.

sangaman commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+ * @returns
+ */
+export function hashTapLeaf(script: Buffer): Buffer {
+  // TODO: use multiple byte `size` when script length is >= 253 bytes
```
Yeah that's a good idea. Alternatively I will revisit this inside this PR and just implement the correct logic for larger lengths, it just wasn't particularly easy to do at first pass and I didn't want to get bogged down on it. At least not without introducing a new dependency on something like https://www.npmjs.com/package/structjs.

Here's the reference logic from the optech workshop, fyi:

```python
def ser_compact_size(l):
    r = b""
    if l < 253:
        r = struct.pack("B", l)
    elif l < 0x10000:
        r = struct.pack("<BH", 253, l)
    elif l < 0x100000000:
        r = struct.pack("<BI", 254, l)
    else:
        r = struct.pack("<BQ", 255, l)
    return r
```

OttoAllmendinger replied:
`varunit-bitcoin` https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/package.json#L65

sangaman replied:
Thanks otto, that looks like just what I need. I'm going to split this logic into its own function and add some test cases.

sangaman commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+      leftChild,
+      rightChild,
+    ]),
+  );
```
I'm a fan.

sangaman commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,205 @@
+  const weightedScripts: WeightedTapScript[] = [];
+
+  scripts.forEach((script, index) => {
+    const weight = weights ? weights[index] || 1 : 1;
```
I tried exactly this but the compiler would not accept it. I honestly wasn't sure if it was due to an older version of typescript or if doing `?.[]` is just invalid, so I went with what's there currently.

tylerlevine replied:
Fine with me. I think now that you've updated the typescript version it should work but no big deal.

sangaman replied:
Indeed you're right, it works now with the newer typescript. I"ll make this change but put it in the commit that upgrades typescript, so if upstream doesn't want to change typescript we can just leave that commit off.

sangaman replied:
Psych, my IDE thinks this syntax is fine, but running `tsc` even with the latest version still craps out on this line. Back to the old way,.

sangaman commented:
@tylerlevine @brandonblack Thanks for the feedback. I was able to address most of the comments. There are still just a couple of loose ends plus I'd still like to add more tests if possible before wrapping this up.

OttoAllmendinger commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,215 @@
+ * @param pubkeys The list of pub keys to aggregate
+ * @returns a 32 byte Buffer representing the aggregate key
+ */
+export function aggregateMuSigPubkeys(pubkeys: Buffer[]): Buffer {
```
assert length > 0 (or maybe > 1 even?)

sangaman replied:
Yes, I think > 1 makes most sense actually.

OttoAllmendinger commented on types/payments/index.d.ts:
```diff
@@ -22,6 +23,8 @@ export interface Payment {
     hash?: Buffer;
     redeem?: Payment;
     witness?: Buffer[];
+    scripts?: Buffer[];
+    weights?: number[];
```
for the p2sh payments for instance, the construction happens via the `output` or `redeem` property: https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/ts_src/payments/p2sh.ts#L91-L94

maybe we should declare `output?: Buffer | Buffer[]` / `redeem?: Payment | Payment[]` instead here to stay consistent

sangaman replied:
I see how we'd (eventually) want to construct output from redeem, but I'm a bit confused by what you mean here. Why could `output` be `Buffer[]`? Wouldn't it always be the 34 byte locking script?

OttoAllmendinger replied:
I think you are correct

OttoAllmendinger reviewed (changes requested):
let's add some test vectors:

https://raw.githubusercontent.com/bitcoin-core/qa-assets/main/unit_test_data/script_assets_test.json

(usage:
https://github.com/bitcoin/bitcoin/blob/3820090bd619ac85ab35eff376c03136fe4a9f04/src/test/script_tests.cpp#L1718)

sangaman commented on ts_src/payments/p2tr.ts:
```diff
@@ -0,0 +1,87 @@
+const { bech32m } = require('bech32');
+
+/**
+ * An internal key with unknown discrete logarithm for eliminating keypath spends.
+ * This is equal to SHA256(uncompressedDER(SECP256K1_GENERATOR_POINT)).
```
Yes I was aware of this and meant to revisit it to at least put a TODO here. My main concern was what to do with the randomly generated integer after using it. If we just discard it then we can't prove an unspendable key path, which might be acceptable. Alternatively I could add it as a field on the `Payment` object - that might make the most sense. What do you think?

sangaman replied:
Oh and the reason I didn't perform the hash here is just because the value is provided explicitly in the BIP. I suppose I could actually perform the hash but I figured a clear comment explaining how it was derived is just as good.

OttoAllmendinger replied:
let's keep the hardcoded value and link to the BIP with a TODO

adding more stuff to `Payment` should be avoided if possible

Xekyo commented:
I guess that you are aware of this, but I just wanted to point out that the Optech Taproot Workshop was published in October 2019, and the latest changes to the Taproot specification were made in August 2020. This may mean that the content of the workshop doesn't necessarily reflect the latest specification. I haven't reviewed the changes here, I just saw in a commit message that examples from the workshop were used, so I thought I'd point it out in case it was not known.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,215 @@
+
+  const tweakedPubkeys: Buffer[] = pubkeys.map(pubkey => {
+    const trimmedPubkey = trimFirstByte(pubkey);
+    const c = bcrypto.sha256(Buffer.concat([L, trimmedPubkey]));
```
Did you see https://github.com/ElementsProject/secp256k1-zkp/blob/881b15cb4371d4b58d851575f2eb1832e0dfd5b2/src/modules/musig/musig-spec.mediawiki?
It looks like secp256k1-zkp is going to implement key aggregation based on MuSig2* instead of MuSig2, which would mean using `1` for the exponent of the second key in the sorted list, which would complicate everything here.

sangaman replied:
Thank you, I did not see this. I'd read elsewhere that the key aggregation routine for MuSig1 was the same as MuSig2, but I wasn't aware of MuSig2*. It's easy enough to implement that logic here, just add this special case for the second element, but I'm not sure how that affects the tests and signing procedure. BIP340 doesn't mention it (it doesn't mention MuSig2 at all). I'm inclined to just leave this alone in this PR but keep it in mind and revisit when we have some signing logic to work with and a better idea of the implications.

brandonblack replied:
My only concern is that once we've generated any wallets using one version of pubkey aggregation, we're pretty locked in. For that reason, I think we should discuss with (at least) Jonas before we committing to an aggregation scheme.

sangaman commented:
> I guess that you are aware of this, but I just wanted to point out that the Optech Taproot Workshop was published in October 2019, and the latest changes to the Taproot specification were made in August 2020. This may mean that the content of the workshop doesn't necessarily reflect the latest specification. I haven't reviewed the changes here, I just saw in a commit message that examples from the workshop were used, so I thought I'd point it out in case it was not known.

Thanks, I was aware that the optech workshop is older than the latest changes to the BIPs, but as far as I could tell the examples from the workshop I referenced were in accordance with BIPs 340/341 and the code accompanying them. There have been ongoing updates to the github repo for the optech examples since they were published, so that may be why there weren't any discrepancies I could detect. Still something to keep in mind.

OttoAllmendinger commented on ts_src/payments/p2tr.ts:
```diff
@@ -0,0 +1,87 @@
+  lazy.prop(o, 'address', () => {
+    if (!o.output) return;
+
+    const words = bech32m.toWords(o.output.slice(2));
```
maybe we should sanity check those two bytes / use the existing methods to validate and parse the script

OttoAllmendinger replied:
OTOH the other payment builders don't do that either

sangaman replied:
The other payment builders do have some "extended validation" logic that I left out for now, I figured it will be easier to implement later and didn't want to get held up by it. I figure validating the first two bytes of the output we're provided makes sense to do there.

Alternatively, since I believe the first two bytes are constant, this builder could only accept the 32 byte output witness program and then add the 2 bytes to create the output script. That makes more sense to me actually since it eliminates the possibility of user error, but might make sense as a followup to this PR. Although I could do it here if it makes sense to everyone.

OttoAllmendinger commented on test/taproot.spec.ts:
```diff
@@ -9,100 +9,128 @@ describe('taproot utils', () => {
-    );
+  describe('musig key aggregation', () => {
+    it('aggregates pubkeys', () => {
+      // example taken from bitcoinops taproot workshop musig exercise
```
link pls

sangaman replied:
I'll also note that most of these expected results are derived from the optech taproot workshop example.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,234 @@
+ */
+export function aggregateMuSigPubkeys(pubkeys: Buffer[]): Buffer {
+  assert(
+    pubkeys.length > 1,
```
https://github.com/jonasnick/secp256k1-zkp/blob/0aeaa5dfb19445f845890f3a4502c934550f4548/include/secp256k1_musig.h#L200
>         n_pubkeys: length of pubkeys array. Must be greater than 0.

sangaman replied:
It seems really odd to me to want to go through the extra steps of key aggregation for a single key, I don't see any benefit from that. My inclination is still to disallow it here, even if it's technically doable in secp256k1-zkp. It shouldn't break compatibility afaict.

sangaman commented:
OK, I went ahead and added the logic for MuSig2*, skipping exponentiation of the second key in the sorted list of keys for key aggregation.

I also got some key aggregation examples directly from the code at https://github.com/jonasnick/secp256k1-zkp/tree/musig2 to test against. In the process I added two new tagged hashes to match the logic there, `KeyAgg list` and `KeyAgg coefficient`. The good news is I was able to get those to pass without too many changes, basically just the MuSig2* and tagged hashes described above.

I resolved the review comments that I thought were addressed pretty straightforwardly by my commits. I think the last step is to get another quick pass on reviews and then clean up the commit history & messages to describe the changes since the original commit.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,274 @@
+  Buffer.from('KeyAgg coefficient'),
+);
+
+const TAGGED_HASH_PREFIXES = {
+  TapLeaf: Buffer.concat([
+    TAPLEAF_TAGGED_HASH_PREFIX,
+    TAPLEAF_TAGGED_HASH_PREFIX,
+  ]),
+  TapBranch: Buffer.concat([
+    TAPBRANCH_TAGGED_HASH_PREFIX,
+    TAPBRANCH_TAGGED_HASH_PREFIX,
+  ]),
+  TapTweak: Buffer.concat([
+    TAPTWEAK_TAGGED_HASH_PREFIX,
+    TAPTWEAK_TAGGED_HASH_PREFIX,
+  ]),
+  'KeyAgg list': Buffer.concat([
+    KEYAGG_LIST_TAGGED_HASH_PREFIX,
+    KEYAGG_LIST_TAGGED_HASH_PREFIX,
+  ]),
+  'KeyAgg coefficient': Buffer.concat([
+    KEYAGG_COEFFICIENT_TAGGED_HASH_PREFIX,
+    KEYAGG_COEFFICIENT_TAGGED_HASH_PREFIX,
+  ]),
+};
```
```suggestion
function makeHashPrefix(tag: Buffer): Buffer {
  return Buffer.concat([tag, tag]);
}

const TAGGED_HASH_PREFIXES = {
  TapLeaf: makeHashPrefix(TAPLEAF_TAGGED_HASH_PREFIX),
  TapBranch: makeHashPrefix(TAPBRANCH_TAGGED_HASH_PREFIX),
  TapTweak: makeHashPrefix(TAPTWEAK_TAGGED_HASH_PREFIX),
  'KeyAgg list': makeHashPrefix(KEYAGG_LIST_TAGGED_HASH_PREFIX),
  'KeyAgg coefficient': makeHashPrefix(KEYAGG_COEFFICIENT_TAGGED_HASH_PREFIX),
};
```

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,274 @@
+const EVEN_Y_COORD_PREFIX = new Uint8Array([0x02]);
+const INITIAL_TAPSCRIPT_VERSION = new Uint8Array([0xc0]);
+
+const TAPLEAF_TAGGED_HASH_PREFIX = bcrypto.sha256(Buffer.from('TapLeaf'));
+const TAPBRANCH_TAGGED_HASH_PREFIX = bcrypto.sha256(Buffer.from('TapBranch'));
+const TAPTWEAK_TAGGED_HASH_PREFIX = bcrypto.sha256(Buffer.from('TapTweak'));
+const KEYAGG_LIST_TAGGED_HASH_PREFIX = bcrypto.sha256(
+  Buffer.from('KeyAgg list'),
+);
+const KEYAGG_COEFFICIENT_TAGGED_HASH_PREFIX = bcrypto.sha256(
+  Buffer.from('KeyAgg coefficient'),
+);
+
+const TAGGED_HASH_PREFIXES = {
+  TapLeaf: Buffer.concat([
+    TAPLEAF_TAGGED_HASH_PREFIX,
+    TAPLEAF_TAGGED_HASH_PREFIX,
+  ]),
+  TapBranch: Buffer.concat([
+    TAPBRANCH_TAGGED_HASH_PREFIX,
+    TAPBRANCH_TAGGED_HASH_PREFIX,
+  ]),
+  TapTweak: Buffer.concat([
+    TAPTWEAK_TAGGED_HASH_PREFIX,
+    TAPTWEAK_TAGGED_HASH_PREFIX,
+  ]),
+  'KeyAgg list': Buffer.concat([
+    KEYAGG_LIST_TAGGED_HASH_PREFIX,
+    KEYAGG_LIST_TAGGED_HASH_PREFIX,
+  ]),
+  'KeyAgg coefficient': Buffer.concat([
+    KEYAGG_COEFFICIENT_TAGGED_HASH_PREFIX,
+    KEYAGG_COEFFICIENT_TAGGED_HASH_PREFIX,
+  ]),
+};
```
```suggestion
const TAGS = [ 'TapLeaf', 'TapBranch', 'TapTweak', 'KeyAgg list', 'KeyAgg coefficient' ];

const TAGGED_HASH_PREFIXES = Object.fromEntries(TAGS.map(tag => {
    const tagBuf = Buffer.from(tag);
    return [tag, Buffer.concat([tag, tag])];
}));
```

brandonblack replied:
Not sure if typescript compiler will handle this, might need to go to an enumerated type, but I haven't done those yet.

sangaman replied:
I like this a lot, particularly since this list of tagged hashes is going to keep growing, and it satisfies the type checking if you define `TAGS` with `as const`, but that (and `Object.fromEntries` I believe) requires typescript 3.4+. I'm inclined to just upgrade the version, if upstream doesn't want to do that we can switch back to what I have here currently.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,274 @@
+
+  const tweakedPubkeys: Buffer[] = trimmedPubkeys.map(
+    (trimmedPubkey, index) => {
+      const pubkey = Buffer.concat([EVEN_Y_COORD_PREFIX, trimmedPubkey]);
```
Love this vs. re-trimming like it used to be. Nice.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,274 @@
+
+  // The aggregate key must be a point with an even y coordinate, so we manually
+  // set the leading byte to 0x02 if it isn't already.
+  return aggregatePubkey.slice(1).compare(EVEN_Y_COORD_PREFIX) === 0
```
I think this should be `.slice(0,1)`, although it might make sense not to branch and just always replace the first byte.

sangaman replied:
Yep thanks for catching this.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,274 @@
+  // set the leading byte to 0x02 if it isn't already.
+  return aggregatePubkey.slice(1).compare(EVEN_Y_COORD_PREFIX) === 0
+    ? aggregatePubkey
+    : Buffer.concat([EVEN_Y_COORD_PREFIX, trimFirstByte(aggregatePubkey)]);
```
Can this be replaced with `aggregatePubkey[0] = EVEN_Y_COORD_PREFIX;`?

sangaman replied:
Yep makes sense, just skip the `slice` and branch.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,256 @@
+      return ecc.pointMultiply(pubkey, c);
+    },
+  );
+  const aggregatePubkey = tweakedPubkeys
+    .slice(1)
+    .reduce((prev, curr) => ecc.pointAdd(prev, curr), tweakedPubkeys[0]);
```
> initialValue Optional
>    A value to which previousValue is initialized the first time the callback is called. If initialValue is specified, that also causes currentValue to be initialized to the first value in the array. If initialValue is not specified, previousValue is initialized to the first value in the array, and currentValue is initialized to the second value in the array.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce
This can be simplified to
```suggestion
const aggregatePubkey = tweakedPubkeys.reduce((prev, cur) => ecc.pointAdd(prev, curr));
```

sangaman replied:
Yep that checks out, the slice is superfluous.

brandonblack replied:
Yeah, I suggested the slice, but I'm a bit rusty still, sorry! :)

tylerlevine commented on ts_src/taproot.ts:
```diff
@@ -0,0 +1,256 @@
+    queue.add(weightedScript);
+  });
+
+  // Now that we have a queue of weighted scripts, we begin a loop whereby we
```
Excellent comments here 👍

tylerlevine reviewed (approved):
awesome work, looks good to me 🚀
